File: app/src/main/java/ru/nsu/fit/modao/fragments/SelectParticipantsFragment.java
package ru.nsu.fit.modao.fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavDirections;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.LinkedList;
import java.util.List;

import ru.nsu.fit.modao.R;
import ru.nsu.fit.modao.adapter.ParticipantsAdapter;
import ru.nsu.fit.modao.databinding.FragmentSelectParticipantsBinding;
import ru.nsu.fit.modao.model.CreateExpenseParticipant;
import ru.nsu.fit.modao.model.User;


public class SelectParticipantsFragment extends Fragment {

    FragmentSelectParticipantsBinding binding;
    List<CreateExpenseParticipant> participants;
    CreateExpenseParticipant sponsor;
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentSelectParticipantsBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        participants  = new LinkedList<>();
        sponsor = SelectParticipantsFragmentArgs.fromBundle(getArguments()).getSponsor();
        participants.add(new CreateExpenseParticipant("Nikita", 1f, 1L, "1"));
        participants.add(new CreateExpenseParticipant("Efim", 1f,2L, "1"));
        participants.add(new CreateExpenseParticipant("Olga", 1f,3L, "1"));
        participants.add(new CreateExpenseParticipant("Peter", 1f,4L, "1"));
        ParticipantsAdapter adapter = new ParticipantsAdapter(getContext(), participants);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getContext(),
                RecyclerView.VERTICAL, false);
        binding.participantRecycler.setLayoutManager(layoutManager);
        binding.participantRecycler.setAdapter(adapter);
        binding.buttonNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                List<CreateExpenseParticipant> list = adapter.getSelected();
                CreateExpenseParticipant[] users = new CreateExpenseParticipant[list.size()];
                list.toArray(users);
                NavDirections action = SelectParticipantsFragmentDirections
                        .actionSelectParticipantsFragmentToEnterCoefficientsFragment(users,
                                SelectParticipantsFragmentArgs.fromBundle(getArguments()).getCost(),
                                sponsor);
                Navigation.findNavController(view).navigate(action);
            }
        });
    }
}

File: app/src/main/java/ru/nsu/fit/modao/fragments/SelectParticipantsFragment.kt
package ru.nsu.fit.modao.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dagger.hilt.android.AndroidEntryPoint
import ru.nsu.fit.modao.adapter.SelectAdapter
import ru.nsu.fit.modao.databinding.FragmentSelectParticipantsBinding
import ru.nsu.fit.modao.models.ParticipantEvent
import ru.nsu.fit.modao.viewmodels.CreateExpenseViewModel
@AndroidEntryPoint
class SelectParticipantsFragment: Fragment(), SelectAdapter.CustomListener {
    private var _binding: FragmentSelectParticipantsBinding? = null
    private val binding get() = _binding!!
    private val createExpenseViewModel: CreateExpenseViewModel by viewModels({requireParentFragment().requireParentFragment()})
    private val adapter = SelectAdapter()
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentSelectParticipantsBinding.inflate(inflater, container, false)
        return binding.root
    }
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        adapter.attachListener(this)
        binding.participantRecycler.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        binding.participantRecycler.adapter = adapter

        createExpenseViewModel.participants.observe(viewLifecycleOwner){
            adapter.setList(it)
        }
        binding.buttonSelect.setOnClickListener {
            adapter.selectAllParticipants()
        }
        binding.buttonUnselect.setOnClickListener {
            adapter.unselectAllParticipants()
        }
    }

    override fun onClickItem(button: CheckBox, participantEvent: ParticipantEvent) {
        participantEvent.selected = button.isChecked
    }
}